### PR TITLE
Schema cleanup and validation

### PR DESCRIPTION
**Pull request by tomlm**
* moved dialogManager into dialogs assembly
* cleanup input schema and comments 
* Add unit tests for validating dialogs/schema (!!!)
* fix a bunch of schema errors
* fix a bunch of dialog errors
* move compnent schema to using botframework-sdk/component.schema
* get rid of sdk.schema, schema comes from SDK assemblies not from sdk.schema
* create tests.schema which is active schema for unit tests
* create testbot.schema which is active schema for testbot

### CODE REVIEW

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/GeneratorTests.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ public static void ClassInitialize(TestContext context)
         {
             ResourceExplorer = new ResourceExplorer()
```
Comment (johnataylor) [inline tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/GeneratorTests.cs:27]: did you intend to remove this file?

File: tests/Microsoft.Bot.Builder.Dialogs.Declarative.Tests/SchemaMergeTests.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            startInfo.UseShellExecute = false;
+            startInfo.WorkingDirectory = projectPath;
+            startInfo.CreateNoWindow = false;
+            startInfo.RedirectStandardError = true;
+
+            // startInfo.RedirectStandardOutput = true;
+            // string output = process.StandardOutput.ReadToEnd();
+
+            var process = Process.Start(startInfo);
+            string error = process.StandardError.ReadToEnd();
```
Comment (johnataylor) [inline tests/Microsoft.Bot.Builder.Dialogs.Declarative.Tests/SchemaMergeTests.cs:77]: var

File: tests/Microsoft.Bot.Builder.Dialogs.Declarative.Tests/SchemaMergeTests.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+            var json = File.ReadAllText(schemaPath);
+            Schema = JSchema.Parse(json);
+        }
+
+        [DataTestMethod]
+        [DynamicData(nameof(Dialogs))]
+        public async Task TestDialogResourcesAreValidForSchema(IResource resource)
+        {
+            FileResource fileResource = resource as FileResource;
```
Comment (johnataylor) [inline tests/Microsoft.Bot.Builder.Dialogs.Declarative.Tests/SchemaMergeTests.cs:88]: var

File: tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs [hunk 4]
```diff
@@ -46,17 +47,17 @@ public MultiplyDialog([CallerFilePath] string sourceFilePath = "", [CallerLineNu
...
         /// </value>
         [JsonProperty("resultProperty")]
-        public string ResultProperty { get; set; }
+        public StringExpression ResultProperty { get; set; }
 
         public override Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
         {
-            var (arg1, err1) = Expression.Parse(Arg1).TryEvaluate(dc.State);
-            var (arg2, err2) = Expression.Parse(Arg2).TryEvaluate(dc.State);
+            var arg1  = Arg1.GetValue(dc.State);
```
Comment (johnataylor) [inline tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs:54]: whitespace picked up by tool

**Comment by chrimc62** [comment]
>         "id": {

How is this id different than the $id included in every loadable object?

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.BeginDialog.schema:8 in c276987. [](commit_id = c276987e12162ae4abde81d59bfb2c8f528f98e7, deletion_comment = False)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.BeginDialog.schema [hunk 5]
```diff
@@ -22,33 +21,34 @@
...
+            "$ref": "https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/component/definitions.schema#/definitions/objectExpression","title": "Options",
             "description": "One or more options that are passed to the dialog that is called.",
             "additionalProperties": {
                 "type": "string",
                 "title": "Options"
-            }
+            },
+            "examples": [
+                {
+                    "arg1": "=expression"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.BeginDialog.schema:40]: >"arg1" [](start = 20, length = 6)

option1?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.DeleteActivity.schema [hunk 6]
```diff
@@ -15,17 +14,15 @@
             "description": "Optional id for the dialog"
         },
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.DeleteActivity.schema:15]: Same here.  $id.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EditActions.schema [hunk 7]
```diff
@@ -15,25 +15,30 @@
...
+                    "enum": [
+                        "insertActions",
+                        "insertActionsBeforeTags",
+                        "appendActions",
+                        "endSequence",
+                        "replaceSequence"
+                    ]
+                },
+                {
+                    "pattern": "^=(\\S){1}"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EditActions.schema:40]: >"pattern": "^=(\\S){1}" [](start = 20, length = 23)

This does not mean what you want.  Should be: ^=([^=]*)$

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EmitEvent.schema [hunk 8]
```diff
@@ -14,53 +14,48 @@
...
+                        "recognizedIntent",
+                        "unknownIntent",
+                        "actionsStarted",
+                        "actionsSaved",
+                        "actionsEnded",
+                        "actionsResumed"
+                    ]
+                },
+                {
+                    "pattern": "^(?!(beginDialog$|resumeDialog$|repromptDialog$|cancelDialog$|endDialog$|activityReceived$|recognizedIntent$|unknownIntent$|actionsStarted$|actionsSaved$|actionsEnded$|actionsResumed))(\\S){1}.*"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EmitEvent.schema:46]: >(\\S){1}.* [](start = 216, length = 10)

Regex

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.DateTimeInput.schema [hunk 9]
```diff
@@ -10,21 +10,31 @@
...
+                        "=user.birthday"
+                    ]
+                },
                 "outputFormat": {
-                    "$role": "expression",
-                    "type": "string",
+                    "$ref": "https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/component/definitions.schema#/definitions/stringExpression",
                     "title": "Output format",
-                    "description": "Expression to format the datetime output.",
+                    "description": "Expression to use format the output.",
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.DateTimeInput.schema:34]: >Expression to use format the output. [](start = 36, length = 36)

Expression to use for formatting the output.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.DateTimeInput.schema [hunk 10]
```diff
@@ -10,21 +10,31 @@
...
                     "examples": [
                         "=this.value[0].Value"
                     ]
-                },
-                "defaultLocale": {
-                    "$role": "expression",
-                    "type": "string",
-                    "title": "Default locale",
-                    "description": "Default locale.",
-                    "default": "en-us"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.DateTimeInput.schema:27]: This moved to multi-locale fallback?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.InputDialog.json [hunk 11]
```diff
@@ -49,78 +50,58 @@
...
       "title": "Validation expressions",
       "description": "Expression to validate user input.",
-      "examples": [
-        "int(this.value) > 1 && int(this.value) <= 150",
-        "count(this.value) < 300"
-      ]
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/component/definitions.schema#/definitions/condition",
+        "title": "Condition",
+        "description": "Expression which needs to met for the input to be considered valid",
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.InputDialog.json:69]: >to [](start = 47, length = 2)

to be

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.InputDialog.json [hunk 12]
```diff
@@ -49,78 +50,58 @@
...
-        "coalesce(@number, @partySize)"
-      ]
-    },
-    "value": {
-      "$role": "expression",
-      "type": [ "object", "array", "number", "integer", "boolean", "string" ],
-      "title": "Value",
-      "description": "'Property' will be set to the value of this expression unless it evaluates to null.",
-      "examples": [
-        "=@userName"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.InputDialog.json:102]: Did you mean to remove?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.QnAMakerDialog.schema [hunk 13]
```diff
@@ -98,15 +90,22 @@
...
+            "oneOf": [
+                {
+                    "enum": [
+                        "default",
+                        "questionOnly",
+                        "autoSuggestQuestion"
+                    ]
+                },
+                {
+                    "pattern": "^=(\\S){1}"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.QnAMakerDialog.schema:105]: >"^=(\\S){1}" [](start = 31, length = 12)

Lots of these bad patterns.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/Microsoft.AdaptiveCardRecognizer.schema [hunk 14]
```diff
@@ -1,15 +0,0 @@
-{
-    "$schema": "https://raw.githubusercontent.com/microsoft/botbuilder-dotnet/master/schemas/component.schema",
-    "$role": "implements(Microsoft.IRecognizer)",
-    "title": "Cross-trained Recognizer Set",
-    "description": "Recognizer for detecting the value response from an Adaptive Card.",
-    "type": "object",
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/Microsoft.AdaptiveCardRecognizer.schema:6]: Was this removed in the code?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Templates/Microsoft.StaticActivityTemplate.schema [hunk 15]
```diff
@@ -1,17 +1,20 @@
...
     "title": "Microsoft Static Activity Template",
     "description": "This allows you to define a static Activity object",
     "type": "object",
     "properties": {
         "activity": {
-            "type": "object",
+            "$ref": "https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/protocol/botframework.json#/definitions/Activity",
             "title": "Activity",
-            "Description": "A static Activity to used"
+            "Description": "A static Activity to used",
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Templates/Microsoft.StaticActivityTemplate.schema:11]: >used [](start = 49, length = 4)

use.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/TriggerConditions/Microsoft.OnCondition.schema [hunk 16]
```diff
@@ -22,15 +21,18 @@
             }
         },
         "priority": {
-            "type": "integer",
+            "$ref": "https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/component/definitions.schema#/definitions/integerExpression",
             "title": "priority",
-            "description": "Priority expression of rule with 0 being the most important",
-            "$role": "expression"
+            "description": "Priority expression of rule with 0 being the most important"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/TriggerConditions/Microsoft.OnCondition.schema:26]: >important [](start = 78, length = 9)

Periods at the end?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/TriggerConditions/Microsoft.OnInvokeActivity.schema [hunk 17]
```diff
@@ -1,5 +1,5 @@
 {
-    "$schema": "https://raw.githubusercontent.com/microsoft/botbuilder-dotnet/master/schemas/component.schema",
+l    "$schema": "https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/component/component.schema",
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/TriggerConditions/Microsoft.OnInvokeActivity.schema:2]: >l  [](start = 0, length = 2)

Guessing you don't want this.

**Comment by chrimc62** [comment]
Two major things:
1) I don't think your regex patterns are right?
2) I'm confused whether we need id and $id.  Should I change $id to id and remove them globally?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EditActions.schema (hunk 7 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EditActions.schema:40]: I'm pretty sure it does, I tried it in multiple regex explorers and also verifiedin the schema validation that vscode does.  In fact, this was key to finding and discovering places in our .dialog files where it didn't follow the pattern.  What this regex says is that
a. start of the line MUST strt with an =
b. there must be at LEAST one non-whitespace char that follows.
Is there a reason you think it's wrong?  

---
In reply to: [410953123](https://github.com/microsoft/botbuilder-dotnet/pull/3758#discussion_r410953123) [](ancestors = 410953123)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EmitEvent.schema (hunk 8 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.EmitEvent.schema:46]: Again I have tested this extensively.  What this combination does is to say the value needs to be one of the enum OR a string which is not that enum.

This is 100% to drive VSCode.  It gives us the options enumerated via the enum in vscode, but allows other values to be passed as well.

---
In reply to: [410953274](https://github.com/microsoft/botbuilder-dotnet/pull/3758#discussion_r410953274) [](ancestors = 410953274)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.QnAMakerDialog.schema (hunk 13 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Dialogs/Microsoft.QnAMakerDialog.schema:105]: Again, these are all by design and correc.t

---
In reply to: [410954158](https://github.com/microsoft/botbuilder-dotnet/pull/3758#discussion_r410954158) [](ancestors = 410954158)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/Microsoft.AdaptiveCardRecognizer.schema (hunk 14 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/Microsoft.AdaptiveCardRecognizer.schema:6]: YEah schema file with no backing class.

---
In reply to: [410954399](https://github.com/microsoft/botbuilder-dotnet/pull/3758#discussion_r410954399) [](ancestors = 410954399)

**Comment by tomlm** [comment]
>         "id": {

resource.id vs dialog.id

---
In reply to: [616186701](https://github.com/microsoft/botbuilder-dotnet/pull/3758#issuecomment-616186701) [](ancestors = 616186701)

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions/Microsoft.BeginDialog.schema:8 in c276987. [](commit_id = c276987e12162ae4abde81d59bfb2c8f528f98e7, deletion_comment = False)

File: libraries/Microsoft.Bot.Builder.AI.LUIS/LuisRecognizer.cs [hunk 18]
```diff
@@ -23,7 +23,7 @@ namespace Microsoft.Bot.Builder.AI.Luis
     public class LuisRecognizer : ITelemetryRecognizer
     {
         [JsonProperty("$kind")]
-        public const string DeclarativeType = "Microsoft.LuisRecognizer";
+        public const string Kind = "Microsoft.LuisRecognizer";
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.AI.LUIS/LuisRecognizer.cs:26]: Breaking change.

**Comment by cleemullins** [comment]
I suggest we close this PR and break it into smaller PRs that are possible to review. The size of this PR, and the mixture of work, makes it really hard to tell what this is.

File: libraries/Microsoft.Bot.Builder.AI.LUIS/LuisRecognizer.cs (hunk 18 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.AI.LUIS/LuisRecognizer.cs:26]: good catch, changing back.

---
In reply to: [411524785](https://github.com/microsoft/botbuilder-dotnet/pull/3758#discussion_r411524785) [](ancestors = 411524785)

**Pull request merged**